tool/tsh/aws.go

```diff
@@ -0,0 +1,344 @@
+	args = append(args, endpointFlag)
+	args = append(args, bundleFlag)
+	cmd := exec.Command(awsCLIBinaryName, args...)
+
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Stdin = os.Stdin
+	if err := cmd.Run(); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func genAndSetAWSCredentials() (*credentials.Credentials, error) {
+	id := uuid.NewUUID().String()
+	secret := uuid.NewUUID().String()
+	if err := setFakeAWSEnvCredentials(id, secret); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return credentials.NewStaticCredentials(id, secret, ""), nil
+}
+
+func createLocalAWSCLIProxy(cf *CLIConf, tc *client.TeleportClient, cred *credentials.Credentials, localCerts tls.Certificate) (*alpnproxy.LocalProxy, error) {
+	if !tc.ALPNSNIListenerEnabled {
+		return nil, trace.NotFound("remote Teleport Proxy doesn't support AWS CLI access protocol")
+	}
+
+	appCerts, err := loadAWSAPPCertificate(tc)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	address, err := utils.ParseAddr(tc.WebProxyAddr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	listener, err := tls.Listen("tcp", "localhost:0", &tls.Config{
+		Certificates: []tls.Certificate{
+			localCerts,
+		},
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	lp, err := alpnproxy.NewLocalProxy(alpnproxy.LocalProxyConfig{
+		Listener:           listener,
+		RemoteProxyAddr:    tc.WebProxyAddr,
+		Protocol:           alpnproxy.ProtocolAWSCLI,
+		InsecureSkipVerify: cf.InsecureSkipVerify,
+		ParentContext:      cf.Context,
+		SNI:                address.Host(),
+		AWSCredentials:     cred,
+		Certs:              []tls.Certificate{appCerts},
+	})
+	if err != nil {
+		if cerr := listener.Close(); cerr != nil {
+			return nil, trace.NewAggregate(err, cerr)
+		}
+		return nil, trace.Wrap(err)
+	}
+	return lp, nil
+}
+
+func loadAWSAPPCertificate(tc *client.TeleportClient) (tls.Certificate, error) {
+	if tc.CurrentAWSCLIApp == "" {
+		return tls.Certificate{}, trace.NotFound("please login into AWS Console App 'tsh app login' first")
+	}
+	key, err := tc.LocalAgent().GetKey(tc.SiteName, client.WithAppCerts{})
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+	cc, ok := key.AppTLSCerts[tc.CurrentAWSCLIApp]
+	if !ok {
+		return tls.Certificate{}, trace.NotFound("please login into AWS Console App 'tsh app login' first")
+	}
+	cert, err := tls.X509KeyPair(cc, key.Priv)
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+	return cert, nil
+}
+
+func validateARNRole(cf *CLIConf, tc *client.TeleportClient, profile *client.ProfileStatus, arnRole string) error {
+	ok := awsarn.IsARN(arnRole)
+
+	validARNs := make(map[string]struct{})
+	for _, roleName := range profile.Roles {
+		role, err := tc.GetRole(cf.Context, roleName)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		roleARNs := role.GetAWSRoleARNs(types.Allow)
+		for _, v := range roleARNs {
+			if v == arnRole {
+				return nil
+			}
+			validARNs[v] = struct{}{}
+		}
```

Comment:
I think this won't work correctly in some cases. For example:
* If some role ARNs are populated with role templating (from IdP traits), this will likely not account for them.
* This code only checks for allowed ARNs and doesn't account for "deny" rules.

Take a look at `CheckAWSRoleARNs` function of `services.AccessChecker`. If we want to be able to display available ARNs, I would just encode what it returns in the certificate.

Reply:
reworked and right now list of available user roles is taken from user identity.